File: app/src/test/java/com/jbrunton/mymovies/TestSchedulerRule.java
package com.jbrunton.mymovies;

import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

import io.reactivex.android.plugins.RxAndroidPlugins;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.schedulers.TestScheduler;

public class TestSchedulerRule implements TestRule {
    public final TestScheduler TEST_SCHEDULER = new TestScheduler();

    @Override public Statement apply(Statement base, Description description) {
        return new Statement() {
            @Override public void evaluate() throws Throwable {
                RxJavaPlugins.setIoSchedulerHandler(
                        scheduler -> TEST_SCHEDULER);
                RxJavaPlugins.setComputationSchedulerHandler(
                        scheduler -> TEST_SCHEDULER);
                RxJavaPlugins.setNewThreadSchedulerHandler(
                        scheduler -> TEST_SCHEDULER);
                RxAndroidPlugins.setInitMainThreadSchedulerHandler(
                        scheduler -> Schedulers.trampoline());

                try {
                    base.evaluate();
                } finally {
                    RxJavaPlugins.reset();
                    RxAndroidPlugins.reset();
                }
            }
        };
    }
}

File: fixtures/src/main/java/com/jbrunton/fixtures/TestSchedulerRule.kt
package com.jbrunton.fixtures

import io.reactivex.plugins.RxJavaPlugins
import io.reactivex.schedulers.TestScheduler
import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement

class TestSchedulerRule : TestRule {
    val TEST_SCHEDULER = TestScheduler()

    override fun apply(base: Statement, description: Description): Statement {
        return object : Statement() {
            @Throws(Throwable::class)
            override fun evaluate() {
                RxJavaPlugins.setIoSchedulerHandler { scheduler -> TEST_SCHEDULER }
                RxJavaPlugins.setComputationSchedulerHandler { scheduler -> TEST_SCHEDULER }
                RxJavaPlugins.setNewThreadSchedulerHandler { scheduler -> TEST_SCHEDULER }

                try {
                    base.evaluate()
                } finally {
                    RxJavaPlugins.reset()
                }
            }
        }
    }
}